### PR TITLE
feat(auth): otp embedded to magic links

### PR DESCRIPTION
**Pull request by miguelpeixe**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR implements OTP functionality as part of magic links.

image

image

#### OTP

While generating a magic link, an OTP is now part of the token, sharing the same expiration period.

A valid OTP contains 3 values:
 - `code`: Randomly generated 6-digit numeric code to be entered by the user
 - `hash`: 60-character randomly generated string stored as a cookie and also part of the OTP authorization request payload
 - `attempts`: The number of failed authorization requests made to this OTP

The code is tied to the hash and supports up to 5 attempts before self-destructing.

If the system cannot generate an OTP, the standard magic link will be sent.

image

### How to test the changes in this Pull Request:

Confirm written tests pass.

#### Standard magic link

1. While logged in as admin, visit the Users dashboard and on a reader (customer/subscriber) click to "Send authentication link" 
2. Confirm the email is sent without OTP instructions
3. Click on the link on a fresh session and confirm the magic link behaves as expected

#### OTP

1. On a new session, click the Sign In modal, enter a registered reader email and click "Send authorization code"
2. Confirm the new email template containing the code
3. Paste the code and confirm you're signed in

#### Expiration

1. On a new session, click the Sign In modal, enter a registered reader email and click "Send authorization code"
2. Enter a random value and confirm the error message
3. Continue trying random values until you see the expiration message and directed back to the email input

#### Persistence

1. On a new session, click the Sign In modal, enter a registered reader email and click "Send authorization code"
2. Refresh the page, click on "My Account" and confirm you're prompted to paste the code

#### Persistence under frictionless authentication

1. Setup an overlay prompt with a Reader Registration block
2. On a fresh session, enter an existing email on the rendered block
3. Confirm the email contains the code and magic link
4. Click on "My Account" and confirm the state of the form is to paste the OTP code

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/class-magic-link.php [hunk 1]
```diff
@@ -231,6 +265,56 @@ public static function clear_user_tokens( $user ) {
...
+	private static function generate_otp( $user ) {
+		if ( ! self::is_otp_enabled( $user ) ) {
+			return null;
+		}
+		$code = self::get_random_otp_code();
+		$hash = \wp_generate_password( 60, false, false );
+
+		/** This filter is documented in wp-includes/pluggable.php */
+		if ( \apply_filters( 'send_auth_cookies', true ) ) {
+			/** Subtract 1 minute to the OTP hash cookie expiration. */
```
Comment (adekbadek) [inline includes/class-magic-link.php:304]:
```suggestion
			/** Subtract 1 minute from the OTP hash cookie expiration. */
```

Comment (adekbadek) [inline includes/class-magic-link.php:304]: Why should one minute be subtracted?

File: includes/templates/reader-activation-emails/otp.php [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
...n" style="width:100%;"><tbody><tr><td style="direction:ltr;font-size:0px;padding:0;text-align:center;"><!--[if mso | IE]><table role="presentation" border="0" cellpadding="0" cellspacing="0"><tr><td class="" style="vertical-align:top;width:600px;" ><![endif]--><div class="mj-column-per-100 mj-outlook-group-fix" style="font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;"><table border="0" cellpadding="0" cellspacing="0" role="presentation" width="100%"><tbody><tr><td style="vertical-align:top;padding:12px;"><table border="0" cellpadding="0" cellspacing="0" role="presentation" width="100%"><tbody><tr><td align="left" style="font-size:0px;padding:0;word-break:break-word;"><div style="font-family:Georgia;font-size:16px;line-height:1.5;text-align:left;color:#000000;"><p>' . __( 'Or copy and paste the link in your browser:', 'newspack' ) . ' <a href="*MAGIC_LINK_URL*" data-type="URL" data-id="*MAGIC_LINK_URL*">*MAGIC_LINK_URL*</a></p></div></td></tr></tbody></table></td></tr></tbody></table></div><!--[if mso | IE]></td></tr></table><![endif]--></td></tr></tbody></table></div><!--[if mso | IE]></td></tr></table><![endif]--></div></body></html>',
```
Comment (adekbadek) [inline includes/templates/reader-activation-emails/otp.php:197]: There's some un-placeholdered content in the HTML payload (`https://newspack.sa.ngrok.io…`)

**Review by adekbadek — CHANGES_REQUESTED** [review]

File: includes/class-magic-link.php (hunk 1 above)
Comment (miguelpeixe) [reply includes/class-magic-link.php]: The OTP hash is attached to the token, so once the token expires, the OTP hash is also invalid.

The existence of an OTP hash cookie determines the "Sign In" modal UI to enter an OTP code. If the token is about to expire, it's best that the OTP hash is no longer available so the reader doesn't have a chance to attempt a request that is guaranteed to fail.

File: includes/templates/reader-activation-emails/otp.php (hunk 2 above)
Comment (miguelpeixe) [reply includes/templates/reader-activation-emails/otp.php]: Oops, thank you for catching that! Fixed by 3b70386db2ad8064e9bb916f4494d895ca3838cf

**Review by adekbadek — APPROVED** [review]

**Comment by miguelpeixe** [comment]
Thanks for the review, @adekbadek!

@thomasguillot when you're around, I'd love your take on the email and the OTP code input styles!

**Pull request merged**